pkg/runner/expression.go

```diff
@@ -1,655 +1,257 @@
+		strategy["fail-fast"] = job.Strategy.FailFast
+		strategy["max-parallel"] = job.Strategy.MaxParallel
 	}
-	return out, isString
-}
 
-// Rewrite tries to transform any javascript property accessor into its bracket notation.
-// For instance, "object.property" would become "object['property']".
-func (ee *expressionEvaluator) Rewrite(in string) string {
-	var buf strings.Builder
-	r := strings.NewReader(in)
-	for {
-		c, _, err := r.ReadRune()
-		if err == io.EOF {
-			break
-		}
-		//nolint
-		switch {
-		default:
-			buf.WriteRune(c)
-		case c == '\'':
-			buf.WriteRune(c)
-			ee.advString(&buf, r)
-		case c == '.':
-			buf.WriteString("['")
-			ee.advPropertyName(&buf, r)
-			buf.WriteString("']")
-		}
-	}
-	return buf.String()
-}
+	jobs := rc.Run.Workflow.Jobs
+	jobNeeds := rc.Run.Job().Needs()
 
-func (*expressionEvaluator) advString(w *strings.Builder, r *strings.Reader) error {
-	for {
-		c, _, err := r.ReadRune()
-		if err != nil {
-			return err
-		}
-		if c != '\'' {
-			w.WriteRune(c) //nolint
-			continue
-		}
-
-		// Handles a escaped string: ex. 'It''s ok'
-		c, _, err = r.ReadRune()
-		if err != nil {
-			w.WriteString("'") //nolint
-			return err
-		}
-		if c != '\'' {
-			w.WriteString("'") //nolint
-			if err := r.UnreadRune(); err != nil {
-				return err
-			}
-			break
+	using := make(map[string]map[string]map[string]string)
+	for _, needs := range jobNeeds {
+		using[needs] = map[string]map[string]string{
+			"outputs": jobs[needs].Outputs,
 		}
-		w.WriteString(`\'`) //nolint
 	}
-	return nil
-}
 
-func (*expressionEvaluator) advPropertyName(w *strings.Builder, r *strings.Reader) error {
-	for {
-		c, _, err := r.ReadRune()
-		if err != nil {
-			return err
-		}
-		if !isLetter(c) {
-			if err := r.UnreadRune(); err != nil {
-				return err
-			}
-			break
-		}
-		w.WriteRune(c) //nolint
+	ee := &exprparser.EvaluationEnvironment{
+		Github: rc.getGithubContext(),
+		Env:    rc.GetEnv(),
+		Job:    rc.getJobContext(),
+		// todo: should be unavailable
+		// but required to interpolate/evaluate the step outputs on the job
+		Steps: rc.getStepsContext(),
+		Runner: map[string]interface{}{
+			"os":         "Linux",
+			"temp":       "/tmp",
+			"tool_cache": "/opt/hostedtoolcache",
+		},
+		Secrets:  rc.Config.Secrets,
+		Strategy: strategy,
+		Matrix:   rc.Matrix,
+		Needs:    using,
+	}
+	return expressionEvaluator{
+		interpreter: exprparser.NewInterpeter(ee, exprparser.Config{
+			Run:        rc.Run,
+			WorkingDir: rc.Config.Workdir,
+			Context:    "job",
+		}),
 	}
-	return nil
 }
 
-func isLetter(c rune) bool {
-	switch {
-	case c >= 'a' && c <= 'z':
-		return true
-	case c >= 'A' && c <= 'Z':
-		return true
-	case c >= '0' && c <= '9':
-		return true
-	case c == '_' || c == '-':
-		return true
-	default:
-		return false
+// NewExpressionEvaluator creates a new evaluator
+func (sc *StepContext) NewExpressionEvaluator() ExpressionEvaluator {
+	rc := sc.RunContext
+	// todo: cleanup EvaluationEnvironment creation
```

Comment:
missed?

Reply:
@ZauberNerd Do you still see a reason for this todo?

Reply:
I don't think there's something urgent to do here. I mean we could try to consolidate some things between StepContext and RunContext, but for now it works, so I'd rather see it merged and if need be addressed in a separate PR.